src/main/java/net/imglib2/img/basictypeaccess/nio/AbstractBufferAccess.java

```diff
@@ -0,0 +1,223 @@
+ * {@link Buffer}.
+ * 
+ * @author Mark Kittisopikul
+ * @author Philipp Hanslovsky (initial specification)
+ *
+ * @param <A> Subclass used as return type for 
+ * {@link #newInstance(Buffer, boolean)} and
+ * {@link #newInstance(ByteBuffer, boolean)}
+ * 
+ * @param <B> {@link Buffer} subclass
+ */
+public abstract class AbstractBufferAccess <A extends AbstractBufferAccess< A, B >, B extends Buffer> implements BufferAccess< A > {
+	
+	/**
+	 * Default valid setting if not specified
+	 */
+	protected static boolean DEFAULT_IS_VALID = true;
+	
+	/**
+	 * If valid or not as per {@link VolatileAccess}.
+	 */
+	protected final boolean isValid;
+
+	protected final B buffer;
+	
+	/*
+	 * Constructors
+	 */
+
+	public AbstractBufferAccess(final B buffer, final boolean isValid) {
+		this.buffer = buffer;
+		this.isValid = isValid;
+	}
+	
+	public AbstractBufferAccess(final B buffer) {
+		this(buffer, DEFAULT_IS_VALID);
+	}
+	
+	public AbstractBufferAccess(final Function<Integer,B> allocate, final int numEntities) {
+		this( allocate.apply(numEntities) );
+	}
+	
+	/*
+	 * Public methods
+	 */
+	
+	/**
+	 * Returns a new instance of BufferAccess with duplicated Buffer for
+	 * thread safety.
+	 */
+	@Override
+	public A createView(Object o) {
+		return newInstance( duplicateBuffer(buffer), isValid() );
+	}
+	
+	@Override
+	public boolean isValid()
+	{
+		return isValid;
+	}
+	
+	/**
+	 * Respects the directness of the buffer.
+	 * If this buffer is direct, then the new backing buffer is also direct.
+	 */
+	@Override
+	public A createArray( int numEntities )
+	{
+		return allocate( numEntities );
+	}
+
+	/**
+	 * Returns the raw buffer (not duplicated)
+	 */
+	@Override
+	public B getCurrentStorageArray()
+	{
+		return buffer;
+	}
+
+	/**
+	 * Returns the Buffer's limit or zero if buffer is null.
+	 * 
+	 * @see Buffer#limit()
+	 */
+	@Override
+	public int getArrayLength()
+	{
+		if( buffer == null )
+			return 0;
+		else
+			return buffer.limit();
+	}
+	
+	/**
+	 * Return the Buffer's directness or false is the buffer is null.
+	 */
+	@Override
+	public boolean isDirect() {
+		return buffer != null && buffer.isDirect();
+	}
+	
+	@Override
+	public boolean isReadOnly() {
+		return buffer.isReadOnly();
+	}
+	
+	/*
+	 * Protected methods
+	 */
+	
+	/**
+	 * Allocate a new ByteBuffer with initial capacity and directness.
+	 * 
+	 * @param numEntities
+	 * @param isDirect
+	 * @return
+	 */
+	protected ByteBuffer allocateByteBuffer( int numEntities, boolean isDirect ) {
+		if(isDirect)
+			return ByteBuffer.allocateDirect( numEntities * getNumBytes() );
+		else
+			return ByteBuffer.allocate( numEntities * getNumBytes() );
+	}
+	
+	/**
+	 * Allocate a new BufferAccess specifying the directness and validity.
+	 * 
+	 * @param numEntities
+	 * @param isDirect
+	 * @param isValid
+	 * @return
+	 */
+	protected A allocate( int numEntities, boolean isDirect, boolean isValid) {
+		return newInstance( allocateByteBuffer(numEntities, isDirect), isValid );
```

Comment:
Hard-coding this to `ByteBuffer` might be an issue:
 1. `newInstance` is generic wrt to the buffer. What happens if we pass a `ByteBuffer` but the generic bound `B` is `DoubleBuffer`?
 2. For any type that has multiple bytes per entity, `numEntities` is restricted to about `Integer.MAX_VALUE / getNumBytes()`. For `double`, that would mean only 1/8 of the capacity compared to `byte`.

Reply:
1. There are currently two `newInstance` methods. One is specified by the interface which is only generic to the `Access` type. Another is specified by `AbstractBufferAccess` which is also generic to the `java.nio.Buffer` implementation.

```java
public A newInstance( ByteBuffer buffer, boolean isValid ); // specified by BufferAccess < A >

public abstract A newInstance(B buffer, boolean isValid ); // specified by AbstractBufferAccess <A , B >
```

The idea is that every `BufferAccess` can either be created by either providing a `ByteBuffer` or a more specific `Buffer` like `LongBuffer`. If given a `ByteBuffer`, then the `ByteBuffer` is converted into the type-specific buffer (e.g. `LongBuffer`)

2. This machinery exist mainly to support `ArrayDataAccess.createArray( int numEntities )`. There is a Java limitation in the language itself in that currently you can only create a _direct_ `Buffer` by first creating a `ByteBuffer` and then transforming it into a more specific `Buffer`. Only `ByteBuffer` has `allocateDirect` as you pointed out earlier ( https://github.com/imglib/imglib2/pull/278#issue-353261538 ).

For a _non-direct_ `Buffer` you could get around this by just constructing a `BufferAccess` by passing a `LongBuffer`.

The solution then is that concrete implementations of `AbstractBufferAccess` should overload `allocate` to create a _non-direct_ of a longer size if available. This is already done by the concrete class constructors. I'll revise this in the concrete classes to make longer buffers available to `ArrayDataAccess.createArray( int numEntities )`, but I think I'll keep this `AbstractBufferAccess` implementation here.

Reply:
My first point is resolved, I did not see the `newInstance( ByteBuffer, boolean )` signature, should've opened in an IDE instead of just looking at the diff. As to my second point, how about adding an abstract method
```java
protected abstract B allocateBuffer( int numEntities, boolean isDirect );
```
and then change `allocate` to
```java
protected A allocate( int numEntities, boolean isDirect, boolean isValid) {
	return newInstance( allocateBuffer(numEntities, isDirect), isValid );
}
```

In the concrete implementations, e.g. `DoubleBuffer`, implement `allocateBuffer` like this:
```java
@Override
protected DoubleBuffer allocateBuffer(int numEntities, boolean isDirect) {
	if(isDirect)
		return ByteBuffer.allocateDirect( numEntities * getNumBytes() ).asDoubleBuffer();
	else
		return DoubleBuffer.allocate( numEntities );
}
```

Then, at least for non-direct buffers, we can use the maximum `numEntitites`/capacity as expected. For the `isDirect` branch, maybe we should add a check and throw an exception if `numEntitites` is too large.

Reply:
Fixed in 83c61ea